kafka/integration_test.go

```diff
@@ -0,0 +1,877 @@
+	msgcnt int64
+	errcnt int64 // count of failed messages
+	msgs   []*Message
+}
+
+// msgtrackerStart sets up a new message tracker
+func msgtrackerStart(t *testing.T, expectedCnt int) (mt msgtracker) {
+	mt = msgtracker{t: t}
+	mt.msgs = make([]*Message, expectedCnt)
+	return mt
+}
+
+var testMsgsInit  = false
+var p0TestMsgs []*testmsgType // partition 0 test messages
+// pAllTestMsgs holds messages for various partitions including PartitionAny and  invalid partitions
+var pAllTestMsgs []*testmsgType
+
+// createTestMessages populates p0TestMsgs and pAllTestMsgs
+func createTestMessages() {
+
+	if testMsgsInit {
+		return
+	}
+	defer func() { testMsgsInit = true }()
+
+	testmsgs := make([]*testmsgType, 100)
+	i := 0
+
+	/*
+		// a test message with default initialization
+		testmsgs[i] = &testmsgType{}
+		i++
+	*/
+
+	// a test message for partition 0 with only Opaque specified
+	testmsgs[i] = &testmsgType{msg: Message{TopicPartition: TopicPartition{Topic: &testconf.Topic, Partition: 0},
+		Opaque: fmt.Sprintf("Op%d", i),
+	}}
+	i++
+
+	// a test message for partition 0 with empty Value and Keys
+	testmsgs[i] = &testmsgType{msg: Message{TopicPartition: TopicPartition{Topic: &testconf.Topic, Partition: 0},
+		Value:  []byte(""),
+		Key:    []byte(""),
+		Opaque: fmt.Sprintf("Op%d", i),
+	}}
+	i++
+
+	// a test message for partition 0 with Value, Key, and Opaque
+	testmsgs[i] = &testmsgType{msg: Message{TopicPartition: TopicPartition{Topic: &testconf.Topic, Partition: 0},
+		Value:  []byte(fmt.Sprintf("value%d", i)),
+		Key:    []byte(fmt.Sprintf("key%d", i)),
+		Opaque: fmt.Sprintf("Op%d", i),
+	}}
+	i++
+
+	// a test message for partition 0 without  Value
+	testmsgs[i] = &testmsgType{msg: Message{TopicPartition: TopicPartition{Topic: &testconf.Topic, Partition: 0},
+		Key:    []byte(fmt.Sprintf("key%d", i)),
+		Opaque: fmt.Sprintf("Op%d", i),
+	}}
+	i++
+
+	// a test message for partition 0 without Key
+	testmsgs[i] = &testmsgType{msg: Message{TopicPartition: TopicPartition{Topic: &testconf.Topic, Partition: 0},
+		Value:  []byte(fmt.Sprintf("value%d", i)),
+		Opaque: fmt.Sprintf("Op%d", i),
+	}}
+	i++
+
+	p0TestMsgs = testmsgs[:i]
+
+	// a test message for PartitonAny with Value, Key, and Opaque
+	testmsgs[i] = &testmsgType{msg: Message{TopicPartition: TopicPartition{Topic: &testconf.Topic, Partition: PartitionAny},
+		Value:  []byte(fmt.Sprintf("value%d", i)),
+		Key:    []byte(fmt.Sprintf("key%d", i)),
+		Opaque: fmt.Sprintf("Op%d", i),
+	}}
+	i++
+
+	// a test message for a non-existent partition with Value, Key, and Opaque.
+	// It should generate ErrUnknownPartition
+	testmsgs[i] = &testmsgType{expectedError: Error{ErrUnknownPartition, ""},
+		msg: Message{TopicPartition: TopicPartition{Topic: &testconf.Topic, Partition: int32(10000)},
+			Value:  []byte(fmt.Sprintf("value%d", i)),
+			Key:    []byte(fmt.Sprintf("key%d", i)),
+			Opaque: fmt.Sprintf("Op%d", i),
+		}}
+	i++
+
+	pAllTestMsgs = testmsgs[:i]
+}
+
+//Test LibraryVersion()
```

Comment:
Some of these seem like normal unit tests, maybe they should be relocated to the relevant `_test.go` files?